Content.Shared/NPC/Systems/NpcTimedFactionSystem.cs

```diff
@@ -0,0 +1,74 @@
+using Content.Shared.NPC.Components;
+using Content.Shared.Timing;
+using Robust.Shared.Random;
+using Robust.Shared.Timing;
+using Content.Shared.NPC.Prototypes;
+
+namespace Content.Shared.NPC.Systems;
+
+/// <summary>
+/// Handles <see cref="NpcTimedFactionComponent"/> faction adding and removal.
+/// </summary>
+public sealed class NpcTimedFactionSystem : EntitySystem
+{
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly NpcFactionSystem _faction = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NpcTimedFactionComponent, MapInitEvent>(OnNpcTimedFactionMapInit);
+        SubscribeLocalEvent<NpcTimedFactionComponent, NpcFactionSystem.TryRemoveFactionAttemptEvent>(OnTryRemoveFaction);
+
+    }
+
+    private void OnNpcTimedFactionMapInit(EntityUid uid, NpcTimedFactionComponent component, MapInitEvent args)
+    {
+        component.TimeFactionChange = _timing.CurTime + component.TimeUntilFactionChange;
+        component.TimeFactionChangeBack = component.TimeFactionChange + component.TimeAsFaction;
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var timedFactionQuery = EntityQueryEnumerator<NpcTimedFactionComponent>();
+        while (timedFactionQuery.MoveNext(out var uid, out var component))
+        {
+            EnsureComp<NpcFactionMemberComponent>(uid, out var facComp);
+            var faction = (uid, facComp);
+
+            if (_timing.CurTime > component.TimeFactionChange)
+            {
+                _faction.AddFaction(faction, component.Faction);
+                component.TimeFactionChange = component.TimeFactionChangeBack + component.TimeUntilFactionChange + TimeSpan.FromSeconds(_random.Next(component.RandomBonusTimeUntilFactionChange + 1));
+
+                if (!component.HasChangedOnce)
+                {
+                    component.HasChangedOnce = true; // Used to prevent event cancellations before the component has triggered.
+                }
+            }
```

Comment:
Wouldn't a refresh event completely negate this entire issue instead of a bunch of custom and messy state handling? See for example the state of metabolism modifiers vs movement speed modifiers.